sssd_test_framework/utils/adcli.py

```diff
@@ -0,0 +1,168 @@
+    for clean up.
+    """
+
+    def info(self, domain: str, *, args: list[str] | None = None) -> ProcessResult:
+        """
+        Discover AD domain.
+
+        :param domain: domain.
+        :type domain: str,
+        :param args: additional arguments, defaults to None
+        :type args: list[str] | None, optional
+        """
+        if args is None:
+            args = []
+
+        return self.host.conn.exec(["adcli", "info", *args, "--verbose", f"{domain}"])
+
+    def testjoin(self, *, args: list[str] | None = None) -> ProcessResult:
+        """
+        Validate join.
+
+        :param args: additional arguments, defaults to None
+        :type args: list[str] | None, optional
+        """
+        if args is None:
+            args = []
+
+        return self.host.conn.exec(["adcli", *args, "--verbose", "testjoin"])
+
+    def delete_computer(
+        self,
+        host: str,
+        *,
+        args: list[str] | None = None,
+        password: str | None = None,
+    ) -> ProcessResult:
+        """
+        Delete computer account.
+
+        :param host: client hostname
+        :type host: str | None,
+        :param args: additional arguments, defaults to None.
+        :type args: list[str] | None, optional
+        :param password: password, defaults to None.
+        :type password: str | None, optional
+        """
+        if args is None:
+            args = []
+
+        return self.host.conn.exec(["adcli", "delete-computer", *args, host], input=password)
+
+    def join(
+        self,
+        domain: str,
+        *,
+        args: list[str] | None = None,
+        password: str | None = None,
+        user: str | None = None,
+        krb: bool | None = False,
+        prompt_password: bool | None = False,
+        stdin: bool | None = False,
+    ) -> ProcessResult:
+        """
+        Create a computer account.
+
+        :param domain: domain.
+        :type domain: str,
+        :param args: additional arguments, defaults to None
+        :type args: list[str] | None, optional
+        :param password: password, defaults to None
+        :type password: str | None, optional
+        :param user: Authenticating User, defaults to None
+        :type user: str | None, optional
+        :param krb: Use kerberos credentials.
+        :type krb: bool | False, optional
+        """
+        if args is None:
+            args = []
+
+        if krb:
+            self.host.conn.exec(["kinit", user], input=password)
+            command = self.host.conn.exec(["adcli", "join", "--verbose", "--login-ccache", *args, domain])
+        elif prompt_password:
+            command = self.host.conn.exec(["adcli", "join", "-W", "--verbose", *args, domain], input=password)
+        elif stdin:
+            command = self.host.conn.exec(
+                ["echo", password, "|", "adcli", "join", "--stdin-password", "--verbose", *args, domain]
+            )
```

Comment:
If krb|prompt_password|stdin are mutually exclusive, is there a better way to either define it or better explanation/description in the docstring?

Reply:
I would love to add that explanation about mutual exclusiveness in docstrings.

Reply:
I've rearranged the logic. Check now.

Reply:
You dropped the --stdin-password option?   If that's ok, then this can be resolved but, if we need that, maybe just a note in the docstring would be enough here.

Reply:
`--stdin-password` and `-W` are doing almost same thing. `--stdin-passwod` is useful with BASH automation, whereas `-W` working with `input=password` is doing same.

Reply:
Do you want a note about `-W` and `--stdin-password` in each method docstring or at class-level note?  I think adding note at each method would be more accessible in long run.

Reply:
I've rewritten the method to have `-W` that is `--prompt-password` as default. Added only condition to check if `kerberos` is `True` or not.

Reply:
Added a note in docstring about `prompt-password` and `--stdin-password`.